### ISSUE TITLE
Viterbi pstate calculations for HMM with initprobs

### ISSUE THREAD

**Issue by bcheggeseth** [chjackson/msm#36]
I'm a fairly new user to `msm` package but I've spent the past few days trying to understand the posterior probability (`pstate`) output from `viterbi.msm()`. I've gone through the mathematics of the Viterbi algorithm. I expect the posterior probability of the first observation per subject to depend on the probability of that outcome as long as `initprobs` argument is provided to `msm` and not equal to `c(1,0,...,0)`. 

Here is a toy example showing posterior probabilities (`pstate` values) from `viterbi.msm()`. 

```
set.seed(123)

sim <- data.frame(time = rep(1:5,50), Id = rep(1:50,each=5))

initialstates <- sample(size = 50, 1:2, prob=c(.5,.5), replace=TRUE)
T <- matrix(c(.8,.1,.2,.9), nrow=2)
states <- matrix(0, nrow=50, ncol = 5)

for(i in 1:50){
  states[i,1] <-  initialstates[i]
  for(j in 2:5){
    states[i,j] <-  sample(size=1,1:2,prob=T[states[i,j-1],],replace=TRUE)
  }
}
  
sim$states <- as.vector(t(states))
m <- c(0,10)
sim$y <- rnorm(50*5,mean = m[sim$states], sd = 1)


hmodel1 <-  list (hmmNorm(mean = 1, sd = 1), hmmNorm(mean= 11, sd = 2))

mod1 <- msm(y ~ time, subject = Id, data = sim, qmatrix = rbind( c(0, 0.01), c(0.01,0)),hmodel=hmodel1)
viterbi.msm(mod1) %>% filter(time == 1) # What you expect given constrained to start in state 1
viterbi.msm(mod1) %>% filter(time == 2) # What you expect based on outcome values



mod2 <- msm(y ~ time, subject = Id, data = sim, qmatrix = rbind( c(0, 0.01), c(0.01,0)),hmodel=hmodel1,initprobs=c(.1,.9))
viterbi.msm(mod2) %>% filter(time == 1) # Issue: Posterior probability should involve the probability of the first outcome value
viterbi.msm(mod2) %>% filter(time == 2) # What you expect based on outcome values
``` 

I'm not an expert in C, but I believe there is a bug in the `Viterbi` function in the `lik.c` file in the `pfwd` calculation when `i`=0 at the beginning of the function and when `obs` = 0 at the end of the `Viterbi()` function. The outcome value is calculated from `GetCensored` but there is no call to`GetOutcomeProb` to be incorporated into `lvold` where the comments say use initprobs.

**Comment by chjackson** [comment]
Thanks for the report and sorry for the hassle.  I think this is now fixed in https://github.com/chjackson/msm/commit/9e89118fa628056162979e4406a76b36b617cdf9 .

**Comment by bcheggeseth** [comment]
The posterior probabilities now match what I was expecting but the fitted values for the initial observations are not reflecting the new posterior probabilities.

**Comment by chjackson** [comment]
Right I see - https://github.com/chjackson/msm/commit/dbc488f56e16e8749f8c1ecef40a7831d1846cfb should now deal with this.  Thanks. 

The Viterbi code must surely be the ugliest code in this package, and needs a complete rewrite.

**Comment by helmingstay** [comment]
Bump - is this issue now resolved?
I've experienced a similar issue aligning output pstates with initial probabilities when using censored observations. 
For reference, I'm using version 1.6.9 (CRAN current).

**Comment by chjackson** [comment]
Yes this bug fix was included in 1.6.9.  Feel free to reopen or open a new issue with a reproducible example if there is still a problem.
